File: sdk/ts-agent-sdk/src/actions.ts
/**
 * Action builder functions for constructing well-formed ActionV1 instances.
 *
 * These mirror the Rust helper constructors in kernel-sdk/src/types.rs.
 */

import type { ActionV1 } from "./types.js";
import { ActionType } from "./types.js";

/** Hex string type (0x-prefixed). */
export type Hex = `0x${string}`;

// ============================================================================
// Internal Helpers
// ============================================================================

/** Convert a hex string to Uint8Array. */
function hexToBytes(hex: Hex): Uint8Array {
  const stripped = hex.slice(2);
  if (stripped.length % 2 !== 0) {
    throw new Error(`Invalid hex string: odd length`);
  }
  const bytes = new Uint8Array(stripped.length / 2);
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = parseInt(stripped.slice(i * 2, i * 2 + 2), 16);
  }
  return bytes;
}

/**
 * Convert a 20-byte EVM address (as hex) to a 32-byte left-padded target.
 *
 * The resulting bytes32 has:
 * - Upper 12 bytes: 0x00
 * - Lower 20 bytes: The EVM address
 */
function addressToBytes32(addr: Hex): Uint8Array {
  const addrBytes = hexToBytes(addr);
  if (addrBytes.length !== 20) {
    throw new Error(`Expected 20-byte address, got ${addrBytes.length} bytes`);
  }
  const result = new Uint8Array(32);
  result.set(addrBytes, 12);
  return result;
}

/**
 * Encode a u256 value (from bigint) in big-endian format for ABI encoding.
 * Values must fit in u128 range (the Rust side uses u128).
 */
function encodeU256BE(value: bigint): Uint8Array {
  if (value < 0n) {
    throw new Error("Value must be non-negative");
  }
  const result = new Uint8Array(32);
  // Write as big-endian: fill from the right
  let v = value;
  for (let i = 31; i >= 0; i--) {
    result[i] = Number(v & 0xffn);
    v >>= 8n;
  }
  return result;
}

// ============================================================================
// Action Builders
// ============================================================================

export const Actions = {
  /**
   * Create a CALL action for on-chain execution.
   *
   * Executed by KernelVault as: `target.call{value: value}(calldata)`
   *
   * @param target - 20-byte EVM address as 0x-prefixed hex
   * @param value - ETH value to send in wei
   * @param calldata - Raw calldata bytes as 0x-prefixed hex
   * @returns ActionV1 with ABI-encoded payload
   *
   * Payload format (ABI-encoded `abi.encode(uint256 value, bytes callData)`):
   * - bytes 0-31: value as uint256 (big-endian)
   * - bytes 32-63: offset to bytes data (always 64 = 0x40)
   * - bytes 64-95: length of callData
   * - bytes 96+: callData (padded to 32-byte boundary)
   */
  call(target: Hex, value: bigint, calldata: Hex): ActionV1 {
    const calldataBytes = hexToBytes(calldata);
    const paddedLen = Math.ceil(calldataBytes.length / 32) * 32;
    const totalSize = 96 + paddedLen;

    const payload = new Uint8Array(totalSize);

    // 1. uint256 value
    payload.set(encodeU256BE(value), 0);

    // 2. offset to bytes data (always 64 = 0x40)
    payload.set(encodeU256BE(64n), 32);

    // 3. length of bytes data
    payload.set(encodeU256BE(BigInt(calldataBytes.length)), 64);

    // 4. bytes data (already zero-padded by Uint8Array initialization)
    payload.set(calldataBytes, 96);

    return {
      actionType: ActionType.CALL,
      target: addressToBytes32(target),
      payload,
    };
  },

  /**
   * Create a TRANSFER_ERC20 action for on-chain execution.
   *
   * Executed by KernelVault as: `IERC20(token).transfer(to, amount)`
   *
   * @param token - ERC20 token address as 0x-prefixed hex
   * @param to - Recipient address as 0x-prefixed hex
   * @param amount - Amount to transfer in token's smallest unit
   * @returns ActionV1 with 96-byte ABI-encoded payload
   *
   * Payload format (ABI-encoded `abi.encode(address token, address to, uint256 amount)`):
   * - bytes 0-31: token address (left-padded to 32 bytes)
   * - bytes 32-63: to address (left-padded to 32 bytes)
   * - bytes 64-95: amount as uint256 (big-endian)
   */
  transferERC20(token: Hex, to: Hex, amount: bigint): ActionV1 {
    const payload = new Uint8Array(96);

    // 1. address token (left-padded)
    payload.set(addressToBytes32(token), 0);

    // 2. address to (left-padded)
    payload.set(addressToBytes32(to), 32);

    // 3. uint256 amount
    payload.set(encodeU256BE(amount), 64);

    return {
      actionType: ActionType.TRANSFER_ERC20,
      target: new Uint8Array(32), // target unused for ERC20 transfers
      payload,
    };
  },

  /**
   * Create a NO_OP action.
   *
   * Skipped during on-chain execution. Useful for padding or placeholders.
   */
  noOp(): ActionV1 {
    return {
      actionType: ActionType.NO_OP,
      target: new Uint8Array(32),
      payload: new Uint8Array(0),
    };
  },
} as const;

File: sdk/py-agent-sdk/tokagent/actions.py
"""Action builder helpers matching the Rust kernel-sdk helpers.

Payload encoding follows Solidity ABI conventions:
- CALL payload: abi.encode(uint256 value, bytes callData)
- TRANSFER_ERC20 payload: abi.encode(address token, address to, uint256 amount)
"""

from __future__ import annotations

from .types import ActionType, ActionV1


# ---------------------------------------------------------------------------
# Address helper
# ---------------------------------------------------------------------------

def address_to_bytes32(addr: str | bytes) -> bytes:
    """Convert a 20-byte EVM address to bytes32 (left-padded with 12 zero bytes).

    Accepts either a hex string (with or without ``0x`` prefix) or raw 20 bytes.
    """
    if isinstance(addr, str):
        addr = bytes.fromhex(addr.removeprefix("0x"))
    if len(addr) != 20:
        raise ValueError(f"address must be 20 bytes, got {len(addr)}")
    return b"\x00" * 12 + addr


# ---------------------------------------------------------------------------
# ABI-encoding helpers (pure stdlib, no eth-abi dependency here)
# ---------------------------------------------------------------------------

def _encode_u256_be(value: int) -> bytes:
    """Encode a non-negative integer as a big-endian uint256 (32 bytes)."""
    return value.to_bytes(32, byteorder="big")


def _encode_call_payload(value: int, call_data: bytes) -> bytes:
    """Encode CALL payload: abi.encode(uint256 value, bytes callData).

    Layout:
      bytes  0-31: value (uint256 big-endian)
      bytes 32-63: offset to bytes data (always 64 = 0x40)
      bytes 64-95: length of callData
      bytes 96+  : callData padded to 32-byte boundary
    """
    data_len = len(call_data)
    padded_len = ((data_len + 31) // 32) * 32

    payload = bytearray()
    payload += _encode_u256_be(value)
    payload += _encode_u256_be(64)           # offset
    payload += _encode_u256_be(data_len)     # length
    payload += call_data
    payload += b"\x00" * (padded_len - data_len)  # padding
    return bytes(payload)


def _encode_transfer_erc20_payload(token: bytes, to: bytes, amount: int) -> bytes:
    """Encode TRANSFER_ERC20 payload: abi.encode(address token, address to, uint256 amount).

    Layout:
      bytes  0-31: token address (left-padded to 32 bytes)
      bytes 32-63: to address (left-padded to 32 bytes)
      bytes 64-95: amount (uint256 big-endian)
    """
    payload = bytearray()
    payload += address_to_bytes32(token)
    payload += address_to_bytes32(to)
    payload += _encode_u256_be(amount)
    return bytes(payload)


# ---------------------------------------------------------------------------
# Action builders
# ---------------------------------------------------------------------------

def call(target: bytes, value: int, calldata: bytes) -> ActionV1:
    """Create a CALL action for on-chain execution.

    Args:
        target: 32-byte target (EVM address left-padded). Use
            ``address_to_bytes32`` to convert a 20-byte address.
        value: ETH value to send in wei.
        calldata: Raw calldata bytes (function selector + encoded args).

    Returns:
        An ``ActionV1`` with type CALL whose payload is ABI-encoded as
        ``abi.encode(uint256 value, bytes callData)``.
    """
    return ActionV1(
        action_type=ActionType.CALL,
        target=target,
        payload=_encode_call_payload(value, calldata),
    )


def transfer_erc20(token: bytes, to: bytes, amount: int) -> ActionV1:
    """Create a TRANSFER_ERC20 action for on-chain execution.

    Args:
        token: 20-byte ERC20 token address.
        to: 20-byte recipient address.
        amount: Amount to transfer (in token's smallest unit).

    Returns:
        An ``ActionV1`` with type TRANSFER_ERC20 whose payload is
        ``abi.encode(address token, address to, uint256 amount)``.
        Target is set to zero bytes (unused for ERC20 transfers).
    """
    return ActionV1(
        action_type=ActionType.TRANSFER_ERC20,
        target=b"\x00" * 32,
        payload=_encode_transfer_erc20_payload(token, to, amount),
    )


def no_op() -> ActionV1:
    """Create a NO_OP action (skipped during on-chain execution)."""
    return ActionV1(
        action_type=ActionType.NO_OP,
        target=b"\x00" * 32,
        payload=b"",
    )
